controllers/admin/AdminCartRulesController.php

```diff
@@ -60,6 +60,96 @@ public function __construct()
         ];
     }
 
+    /**
+     * Function used to render the list to display for this controller.
+     * Add joins to show only cart_rules to which the employee has access, restrict edit and delete action
+     *
+     * @return string|false
+     *
+     * @throws PrestaShopException
+     */
+    public function renderList()
+    {
+        if (!($this->fields_list && is_array($this->fields_list))) {
+            return false;
+        }
+
+        $context = Context::getContext();
+        $all_shops = Shop::getCompleteListOfShopsID();
+        if ($context->employee->isSuperAdmin()) {
+            $employee_shops = $all_shops;
+        } else {
+            $employee_shops = $context->employee->getAssociatedShops();
+        }
+        $context_shops = Shop::getContextListShopID();
+        $employee_has_full_access = empty(array_diff($all_shops, $employee_shops));
+        $this->_select = ($employee_has_full_access ? '0' : '(cse.id_cart_rule is not null or cs.id_cart_rule is null)') . ' as restricted';
+        $this->_join = 'left join `' . _DB_PREFIX_ . 'cart_rule_shop` cs on cs.id_cart_rule = a.id_cart_rule ' . // if there is none, the cart_rule is in all shops
+            'left join `' . _DB_PREFIX_ . 'cart_rule_shop` csc on csc.id_cart_rule = a.id_cart_rule and csc.id_shop in (' . implode(', ', $context_shops) . ') ' . // shops in current context where employee has access
+            'left join `' . _DB_PREFIX_ . 'cart_rule_shop` cse on cse.id_cart_rule = a.id_cart_rule and cse.id_shop not in (' . implode(', ', $employee_shops) . ') '; // if a cart_rule belongs to a shop where the employee has no access then restrict
+        $this->_where = 'and (case when cs.id_cart_rule is null then 1 else (csc.id_cart_rule is not null) end) = 1';
+        $this->_group = 'group by a.id_cart_rule';
+
+        $this->getList($this->context->language->id);
+        $restricted = [];
+        foreach ($this->_list as $row) {
+            if ($row['restricted']) {
+                $restricted[] = $row['id_cart_rule'];
+            }
+        }
+
+        // If list has 'active' field, we automatically create bulk action
+        if (array_key_exists('active', $this->fields_list) && $this->fields_list['active'] === true) {
+            if (!is_array($this->bulk_actions)) {
+                $this->bulk_actions = [];
+            }
+
+            $this->bulk_actions = array_merge([
+                'enableSelection' => [
+                    'text' => $this->trans('Enable selection'),
```

Comment:
I can't see the translation domain in the code. It should be in Admin.Actions, is it? :-)

Reply:
@Julievrz I copied this from `AdminController.php`. Looking closer at it, I see that there are many many strings without domain in `AdminController.php`. They are all translated in `messages.xlf`.

@PierreRambaud also reviewed code I copied from that file and suggested improvements. It would be worth to open a PR for `AdminController.php`. But how should I handle all the other cases of missing domains?

Reply:
Oh. I know some strings are in 'messages' and each time I see one I try to put it in the right domain. Maybe we could continue fixing them as we go when we see them in a pull request?

Reply:
@Julievrz see #28171 for cleaning up translations in `AdminController.php`.

Reply:
Thanks @lmeyer1!